src/main/kotlin/com/openlattice/organizations/ExternalDatabaseManagementService.kt

```diff
@@ -0,0 +1,369 @@
+        }
+        privilegesByUser.forEach { (user, privileges) ->
+            val principal = securePrincipalsManager.getPrincipal(user).principal
+            val aceKey = AceKey(aclKey, principal)
+            var permissions = EnumSet.noneOf(Permission::class.java)
+            if (privileges == OwnerPrivileges.values() as Set<String>) { //eek
+                permissions.add(Permission.OWNER)
+            }
+            if (privileges.contains("SELECT")) {
+                permissions.add(Permission.READ)
+            }
+            if (privileges.contains("INSERT") || privileges.contains("UPDATE"))
+                permissions.add(Permission.WRITE)
+            val aceValue = AceValue(permissions, SecurableObjectType.OrganizationExternalDatabaseTable, OffsetDateTime.MAX)
+            aces[aceKey] = aceValue //TODO should be a merger or what?
+        }
+    }
+
+    /*IDK WHERE THESE BELONG CONCEPTUALLY YET*/
+    fun getColumnNamesByTable(orgId: UUID, dbName: String): Map<String, Set<String>> {
+        val columnNamesByTableName = HashMap<String, MutableSet<String>>()
+        assemblerConnectionManager.connect(dbName).use { dataSource ->
+            dataSource.connection.createStatement().use { stmt ->
+                val rs = stmt.executeQuery(
+                        "SELECT information_schema.tables.table_name, information_schema.columns.column_name " +
+                                "FROM information_schema.tables " +
+                                "LEFT JOIN information_schema.columns on " +
+                                "information_schema.tables.table_name = information_schema.columns.table_name " +
+                                "WHERE information_schema.tables.table_schema='$PUBLIC_SCHEMA' " +
+                                "AND table_type='BASE TABLE' " +
+                                "ORDER BY information_schema.tables.table_name")
+                while (rs.next()) {
+                    val tableName = rs.getString("table_name")
+                    val columnName = rs.getString("column_name")
+                    columnNamesByTableName.getOrPut(tableName) { mutableSetOf() }.add(columnName)
+                }
+            }
+
+        }
+        return columnNamesByTableName
+    }
+
+    fun createNewColumnObjects(dbName: String, tableName: String, tableId: UUID, orgId: UUID, columnName: Optional<String>): Set<OrganizationExternalDatabaseColumn> {
+        val columnCondition = if (columnName.isPresent) {
+            "AND information_schema.columns.column_name = '${columnName.get()}'"
+        } else {
+            ""
+        }
+        val newColumns = mutableSetOf<OrganizationExternalDatabaseColumn>()
+        assemblerConnectionManager.connect(dbName).use { dataSource ->
+            dataSource.connection.createStatement().use { stmt ->
+                val rs = stmt.executeQuery(
+                        "SELECT information_schema.tables.table_name, information_schema.columns.column_name, " +
```

Comment:
This query looks like it has a lot of overlap with the query above -- is there a way to consolidate these?

Reply:
Yea, all of these raw strings I'm planning on abstracting out, just while I'm still writing it helps me to see the queries directly as they still may need to be changed (I haven't tested them yet)